plugins/hls-overloaded-record-dot-plugin/src/Ide/Plugin/OverloadedRecordDot.hs

```diff
@@ -0,0 +1,240 @@
+  pure $ List actions
+  where
+    mkCodeAction :: NormalizedFilePath -> [Extension] -> NextPragmaInfo -> ConvertedRecordSelector -> Command |? CodeAction
+    mkCodeAction nfp exts pragma crs = InR CodeAction
+      { _title = mkCodeActionTitle exts
+      , _kind = Just CodeActionRefactorRewrite
+      , _diagnostics = Nothing
+      , _isPreferred = Nothing
+      , _disabled = Nothing
+      , _edit = Just $ mkWorkspaceEdit nfp edits
+      , _command = Nothing
+      , _xdata = Nothing
+      }
+      where
+        edits = mkTextEdit crs : maybeToList pragmaEdit
+
+        mkTextEdit :: ConvertedRecordSelector -> TextEdit
+        mkTextEdit (ConvertedRecordSelector r cdr) = TextEdit r cdr
+
+        pragmaEdit :: Maybe TextEdit
+        pragmaEdit = if OverloadedRecordDot `elem` exts
+                       then Nothing
+                       else Just $ insertNewPragma pragma OverloadedRecordDot
+
+    mkWorkspaceEdit :: NormalizedFilePath -> [TextEdit] -> WorkspaceEdit
+    mkWorkspaceEdit nfp edits = WorkspaceEdit changes Nothing Nothing
+      where
+        changes = Just $ HashMap.singleton (fromNormalizedUri (normalizedFilePathToUri nfp)) (List edits)
+
+    mkCodeActionTitle :: [Extension] -> Text
+    mkCodeActionTitle exts =
+      if OverloadedRecordDot `elem` exts
+        then title
+        else title <> " (needs extension: OverloadedRecordDot)"
+        where
+          title = "Convert to record dot syntax"
+
+collectConvRecSelsRule :: Recorder (WithPriority Log) -> Rules ()
+collectConvRecSelsRule recorder = define (cmapWithPrio LogShake recorder) $ \CollectRecordSelectors nfp ->
+  use TypeCheck nfp >>= \case
+    Nothing -> pure ([], Nothing)
+    Just tmr -> do
+      let exts = getEnabledExtensions tmr
+          recSels = getRecordSelectors tmr
+      logWith recorder Debug (LogCollectedRecordSelectors recSels)
+      let convertedRecordSelectors = traverse convertRecordSelectors recSels
+          crsMap = RangeMap.fromList range <$> convertedRecordSelectors
+      logWith recorder Debug (LogRenderedRecordSelectors (concat convertedRecordSelectors))
+      pure ([], CCRSR <$> crsMap <*> Just exts)
+  where
+    getEnabledExtensions :: TcModuleResult -> [GhcExtension]
+    getEnabledExtensions = map GhcExtension . getExtensions . tmrParsed
+
+getRecordSelectors :: TcModuleResult -> [RecordSelectors]
+getRecordSelectors (tmrRenamed -> (hs_valds -> valBinds,_,_,_)) =
+    collectRecordSelectors valBinds
+
+convertRecordSelectors :: RecordSelectors -> Maybe ConvertedRecordSelector
+convertRecordSelectors (RecordSelectors ss expr) = ConvertedRecordSelector (realSrcSpanToRange ss) <$> convertRecSel expr
+
+convertRecSel :: Outputable (HsExpr (GhcPass c)) => HsExpr (GhcPass c) -> Maybe Text
+convertRecSel (HsApp _ s r) =
+    Just $ printOutputable r <> "." <> printOutputable s
+convertRecSel ( OpApp _ s _ r) =
+    Just $ "(" <> printOutputable r <> ")." <> printOutputable s
```

Comment:
I think you might want to seriously consider using exactprint. I think this may fall down with some of the more awkward test cases I proposed, like the multi-line ones.

Reply:
I may need to move to exactprint. I ended up using the aptly named "parenthesizeHsExpr" function, and it works. The only problem is that formatting (some spaces) is not kept at the moment.

Reply:
ExactPrint seems like it was built to transform the AST while keeping the original formatting choices, which is exactly what I need. However, it seems like actually using it in this plugin may be unfeasible. As of GHC 9.2, the annotations are provided by GHC, but from what I can gather, `GetParsedModule` doesn't provide all the annotations, requiring the use of `GetParsedModuleWithComments`
https://github.com/haskell/haskell-language-server/blob/437830ae9a5f6832fa391d8365457ad883e898a3/ghcide/src/Development/IDE/Core/Rules.hs#L325-L328
The problem is this plugin uses `TypeCheck`, which depends on `GetParsedModule`, and as far as I know, there is no `TypeCheckedWithComments` rule. Moreover, duplicating the rules in this manner means doubling the compilation time upon each change, which seems unadvisable. It seems like it would be better to only provide rules with annotations, but I assume there is some reason why HLS currently offers both.

In any case, while sometimes the rewriting can lose some of the formatting choices of the user, it still always returns legal syntax. Because of this and the above-mentioned problems with using ExactPrint I feel that printOutputable is currently the only viable option.